pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,333 @@
+		clientv3.OpDelete(key),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	getResp := txnResp.Responses[0].GetResponseRange()
+	if len(getResp.Kvs) == 0 {
+		// To be compatible with original logic of etcdHelper.
+		// In etcd2, delete non-existing key will return error of "key not found"
+		return storage.NewKeyNotFoundError(key, 0)
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) GetToList(ctx context.Context, key string, filter storage.FilterFunc, listObj runtime.Object) error {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+	if len(getResp.Kvs) == 0 {
+		return nil
+	}
+	// Make a list off of the single returned value.
+	values := append([][]byte(nil), getResp.Kvs[0].Value)
+	revs := append([]uint64(nil), uint64(getResp.Kvs[0].ModRevision))
+	if err := decodeList(values, revs, filter, listPtr, h.codec, h.versioner); err != nil {
+		return err
+	}
+	// update version with cluster level revision
+	return h.versioner.UpdateList(listObj, uint64(getResp.Header.Revision))
+}
+
+func (h *v3Helper) List(ctx context.Context, key, resourceVersion string, filter storage.FilterFunc, listObj runtime.Object) error {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	if key[len(key)-1:] != "/" {
+		key += "/"
+	}
+	// Listing all keys that has the prefix of "key"
+	getResp, err := h.client.KV.Get(ctx, key, clientv3.WithPrefix())
+	if err != nil {
+		return err
+	}
+	values := make([][]byte, len(getResp.Kvs))
+	revs := make([]uint64, len(values))
+	for i, kv := range getResp.Kvs {
+		values[i] = kv.Value
+		revs[i] = uint64(kv.ModRevision)
+	}
+	if err := decodeList(values, revs, filter, listPtr, h.codec, h.versioner); err != nil {
+		return err
+	}
+	// update version with cluster level revision
+	return h.versioner.UpdateList(listObj, uint64(getResp.Header.Revision))
+}
+
+func (h *v3Helper) GuaranteedUpdate(ctx context.Context, key string, ptrToType runtime.Object, ignoreNotFound bool, tryUpdate storage.UpdateFunc) error {
+	panic("TODO: unimplemented")
+}
+
+func (h *v3Helper) Watch(ctx context.Context, key string, resourceVersion string, filter storage.FilterFunc) (watch.Interface, error) {
+	panic("TODO: unimplemented")
+}
+
+func (h *v3Helper) WatchList(ctx context.Context, key string, resourceVersion string, filter storage.FilterFunc) (watch.Interface, error) {
+	panic("TODO: unimplemented")
+}
+
+func keyWithPrefix(prefix, key string) string {
+	if strings.HasPrefix(key, prefix) {
+		return key
+	}
+	return path.Join(prefix, key)
+}
+
+// extractRevision extracts revision from obj.
+// If obj has resource version >0, it will be reset to 0.
+func extractRevision(versioner storage.Versioner, obj runtime.Object) (int64, error) {
```

Comment:
We should probably rename it to reflect the fact that resource version is set to 0 if different.
Maybe:
extractAndResetRevision
?